lib/src/importer/node_package.dart

```diff
@@ -0,0 +1,359 @@
+// Copyright 2023 Google Inc. Use of this source code is governed by an
+// MIT-style license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import 'package:collection/collection.dart';
+
+import '../importer.dart';
+import './utils.dart';
+import 'dart:convert';
+import '../io.dart';
+import 'package:path/path.dart' as p;
+
+/// A filesystem importer to use for load implementation details, and for
+/// canonicalizing paths not defined in package.json.
+///
+final _filesystemImporter = FilesystemImporter('.');
+
+/// An importer that resolves `pkg:` URLs using the Node resolution algorithm.
+class NodePackageImporterInternal extends Importer {
+  final Uri entryPointURL;
+  // Creates an importer with the associated entry point url
+  NodePackageImporterInternal(this.entryPointURL);
+
+  @override
+  bool isNonCanonicalScheme(String scheme) {
+    return scheme == 'pkg';
+  }
+
+  @override
+  Uri? canonicalize(Uri url) {
+    if (url.scheme != 'pkg') return null;
+    // TODO(jamesnw) Can these errors even be thrown? Or are these cases
+    // filtered out before this?
+    if (url.path.startsWith('/')) {
+      throw "pkg: URL $url must not be an absolute path.";
+    }
+    if (url.path.isEmpty) {
+      throw "pkg: URL $url must not have an empty path.";
+    }
+    if (url.userInfo != '' || url.hasPort || url.hasQuery || url.hasFragment) {
```

Comment:
@nex3 I'm wondering  if these checks are necessary. My tests to check these cases are failing before `canonicalize` is ever called, but I didn't see in the `modules` spec any specifications on validation before `canonicalize` is called.

I added these checks as they are explicitly named in the [Semantics for all Package Importers](https://github.com/sass/sass/blob/main/proposal/package-importer.d.ts.md#package-importers).

Reply:
Confirmed yes in a meeting, added in ece578e8fcd66ebdf28ce45824bc47e2d031c918.